### DISCUSSION TITLE
Setting the size of a list vector when the children are structs

### DISCUSSION THREAD

**Discussion by prashanthellina** [duckdb/duckdb#14912] [Q&A] [answered]
I use `duckdb_list_vector_set_size()` to set the size of the child vector of a list vector as I add more entries into it. I believe I understand how it works when the list elements are simple types (such as UINTEGER, UTINYINT, VARCHAR etc). However, when child element type is a `STRUCT`, I am not sure setting the list size automatically resizes that member vectors of the struct itself or should I do that explicitly?

UPDATE:
If I have to explicitly set the size of the struct member vectors, I don't see any C-API function for it i.e. something like `duckdb_vector_set_size`

**Comment by Maxxen** [comment]
`duckdb_list_vector_set_size` should only be called on actual list vectors. However, It doesn't set the _capacity_ of the child, for that you need `duckdb_list_vector_reserve`, and you should call that on the list vector too, not the child, e.g. the struct, or its members.

**Reply by Maxxen** [reply]
Have a look at `ListVector::SetSize` and `ListVector::Reserve` in `vector.cpp`.

Usually the pattern is to first reserve enough space in the child vector that you think you will need (or call it again to reserve more space when you need), then in the end run SetSize to tell duckdb how much you actually wrote. 

Note that `ListVector::Reserve` will potentially invalidate pointers into the child vector, so you may need call `duckdb_vector_get_data` again.

**Reply by prashanthellina** [reply]
@Maxxen Thanks for the response. I do already use `duckdb_list_vector_reserve` to ensure there is enough capacity in the list vector and then use `duckdb_list_vector_set_size` to set the actual size (such that size <= reserved_capacity).

However, my confusion is with regard to the struct member vectors when the struct is the list element type. Do I have to reserve capacity and resize them as well or just operating on the list vector itself take care of them?  I understand from your response that you cannot do reserve and set size on non list vectors. Does that mean, DuckDB will internally make sure to adjust the struct member vectors automatically?

**Reply by prashanthellina** [reply]
@Maxxen Please ignore my recent comment above. I will write a code snippet and post to clarify.

**Reply by Maxxen** [reply, marked as answer]
Yes, if a struct list child gets resized, so will all of its members. `void Vector::FindResizeInfos(vector<ResizeInfo> &resize_infos, const idx_t multiplier)` in `vector.cpp` has this logic.

**Reply by prashanthellina** [reply]
Ah! That is good to know. Thank you.

**Reply by prashanthellina** [reply]
> Usually the pattern is to first reserve enough space in the child vector that you think you will need (or call it again to reserve more space when you need), then in the end run SetSize to tell duckdb how much you actually wrote.

This is very useful information. I've been setting the size *before* updating a child element. Updating the size after writing N child elements is much more convenient.
